File: leetcode173.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:
    def __init__(self, root):
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

# @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return len(self.stack) > 0

# @return an integer, the next smallest number
    def next(self):
        node = self.stack.pop()
        x = node.right
        while x:
            self.stack.append(x)
            x = x.left
        return node.val



            # Your BSTIterator object will be instantiated and called as such:
            # obj = BSTIterator(root)
            # param_1 = obj.next()
            # param_2 = obj.hasNext()

File: leetcodec++/leetcode173.cpp
//
// Created by 黄敏辉 on 2019-08-25.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> s;
public:
    BSTIterator(TreeNode* root) {
        if (!root)
            return;
        s.push(root);
        TreeNode* tmp = root->left;
        while (tmp){
            s.push(tmp);
            tmp = tmp->left;
        }
    }

    /** @return the next smallest number */
    int next() {
        TreeNode* n = s.top();
        s.pop();
        if (n->right){
            TreeNode* tmp = n->right;
            while (tmp){
                s.push(tmp);
                tmp = tmp->left;
            }
        }
        return n->val;
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !s.empty();
    }
};